### ISSUE TITLE
508-defect-1 [COGNITION]: The page MUST specify the language

### ISSUE THREAD

**Issue by jenstrickland** [department-of-veterans-affairs/va.gov-team#14432]
# [508-defect-1](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/platform/accessibility/guidance/defect-severity-rubric.md#508-defect-2)

**Feedback framework**

- **❗️ Must** for if the feedback must be applied 
- **⚠️ Should** if the feedback is best practice 
- **✔️ Consider** for suggestions/enhancements 

## Point of Contact

**VFS Point of Contact:** Josh Kim

## User Story or Problem Statement

As a Spanish-speaking assistive tech user, I expect to use my assistive technology in Spanish. 

## Details

~~For the following three pages, the pages are in Spanish and need to indicate that in the markup:~~

~~- [About VA Form 10-10EZ (esp)](https://www.va.gov/find-forms/about-form-10-10ez-espanol/)~~

~~- [About VA Form 10-0137-espanol](https://www.va.gov/find-forms/about-form-10-0137-espanol/)~~

~~- [About VA Form 10-0137 (espanol)](https://www.va.gov/find-forms/about-form-10-0137-espanol/)~~

Update 3/31/2021 - The following eleven detail pages reference Spanish-language forms.  The text should be in Spanish, and the language indicated in the markup. 

| Column1.attributes.form_details_url                          |
|--------------------------------------------------------------|
| https://www.va.gov/find-forms/about-form-10-0137-espanol     |
| https://www.va.gov/find-forms/about-form-10-10ez-espanol     |
| https://www.va.gov/find-forms/about-form-21p-0510-espanol    |
| https://www.va.gov/find-forms/about-form-21p-0512s-1-espanol |
| https://www.va.gov/find-forms/about-form-21p-0513-1-espanol  |
| https://www.va.gov/find-forms/about-form-21p-0514-1-espanol  |
| https://www.va.gov/find-forms/about-form-21p-0516-1-espanol  |
| https://www.va.gov/find-forms/about-form-21p-0517-1-espanol  |
| https://www.va.gov/find-forms/about-form-21p-0518-1-espanol  |
| https://www.va.gov/find-forms/about-form-21p-0519c-1-espanol |
| https://www.va.gov/find-forms/about-form-21p-0519s-1-espanol |

Setting every webpage’s HTML language is an easy way to help your users, including those browsing your website with assistive technology. Setting a language is important because the way that screen readers pronounce words depends on the HTML language assigned to your website.

## Acceptance Criteria

- [ ] The page **MUST** specify the language (`<html lang="en">`) [WCAG 3.1.1](https://www.wuhcag.com/language-of-page/). 

- [ ] Changes in the language within the page **MUST** be specified (e.g. `<span lang="es">Hola</span>`) [WCAG 3.1.2](https://www.wuhcag.com/language-of-parts/).

## Definition of done

1. Review and acknowledge feedback.
1. Fix and/or document decisions made.
1. Accessibility specialist will close ticket after reviewing documented decisions / validating fix.

## Environment

* Operating System: all
* Browser: all
* Screenreading device: all
* Server destination: production

## Steps to Recreate

1. Enter `https://www.va.gov/find-forms/about-form-10-10ez-espanol/` in browser
2. Inspect the page source
3. Verify that the current code is `<html lang="en">` as shown in the screenshot (this indicates the page is in English)
4. Verify that the paragraphs that are in Spanish do not specify the language

## Solution (if known)

The options here are:

- Change the language of the page to Spanish `<html lang="es">`

- Change the areas of the page to indicate those are in Spanish, such as:

```html
<!-- current code -->

<p>Para que los Veteranos soliciten su afiliación al sistema de atención médica de VA. la información proporcionada en este formulario será utilizada por VA para determinar si usted reúne los requisitos para recibir beneficios médicos.</p>`

<!-- recommended code -->
 
<p lang="es">Para que los Veteranos soliciten su afiliación al sistema de atención médica de VA. la información proporcionada en este formulario será utilizada por VA para determinar si usted reúne los requisitos para recibir beneficios médicos.</p>
```

## WCAG or Vendor Guidance (optional)

(WUHCAG.com links provided because they are easier to understand than the W3C WCAG spec pages)
* [WCAG 3.1.1 – Language of Page (Level A, WUHCAG.com)](https://www.wuhcag.com/language-of-page/)

* [WCAG 3.1.2 – Language of Parts (Level AA), WUHCAG.com](https://www.wuhcag.com/language-of-parts/)

* [W3C Declaring language in HTML](https://www.w3.org/International/questions/qa-html-language-declarations)

## Screenshots or Trace Logs

![Screen Shot 2020-10-05 at 5 55 51 PM](https://user-images.githubusercontent.com/57469/95137524-460fc880-0736-11eb-9747-9186320c52ac.png)

![Screen Shot 2020-10-05 at 6 12 43 PM](https://user-images.githubusercontent.com/57469/95137572-6475c400-0736-11eb-9a15-6c4b25e9937e.png)

**Comment by MarciMcGuireGCIO** [comment]
assigning to Decision Tools team; un-assigning from Public Websites

**Comment by zacharymorel** [comment]
Note to self for grooming and planning, this may need a collaboration between teams to discover if we have the data to tell us what the language is.

**Comment by jenstrickland** [comment]
@zacharymorel — if your team needs help determining what language is needed, please tag me. I know many languages. I think this one is self-explanatory as Spanish. Most of the non-English language that VSA teams are adding right now is Spanish.

**Comment by MarciMcGuireGCIO** [comment]
We will add a Discovery spike to this issue to determine what dependencies exist.

**Comment by zacharymorel** [comment]
Thank you, will do @jenstrickland !

**Comment by MarciMcGuireGCIO** [comment]
Related Slack thread: https://dsva.slack.com/archives/CDHBKAL9W/p1609795520311900

**Comment by zacharymorel** [comment]
Another related [thread](https://dsva.slack.com/archives/CUB5X5MGF/p1609794779352100) and graphql [maintained query](https://github.com/department-of-veterans-affairs/vets-website/blob/master/src/site/stages/build/drupal/graphql/vaFormPage.graphql.js#L80-L81) entity for the build process.

**Comment by MarciMcGuireGCIO** [comment]
Closed https://github.com/department-of-veterans-affairs/va.gov-team/issues/1803, which is a duplicate of this issue.

**Comment by zacharymorel** [comment]
See this as well for additional context: https://github.com/department-of-veterans-affairs/va.gov-team/issues/18031

**Comment by MarciMcGuireGCIO** [comment]
Specifics about which info on the page comes from the forms DB: https://dsva.slack.com/files/ULRRWR95K/F018TKDJEHY/image.png

And what is editable in the CMS: 
https://dsva.slack.com/files/ULRRWR95K/F018S8TR9QD/image.png

**Comment by MarciMcGuireGCIO** [comment]
This is blocked while the Content Localization team researches & creates guidance on handling multiple languages.

**Comment by joshkimux** [comment]
@MarciMcGuireGCIO given the recent work with handling multiple languages, is there any chance we could unblock this ticket? @SKasimow for awareness 🙇

**Comment by joshkimux** [comment]
I've updated this to a 508-defect-1 to maintain consistency with other `lang` related tickets. This is a real blocker as Spanish screen reader users may not be able to understand any content pronounced in an English accent.

**Comment by MarciMcGuireGCIO** [comment]
@joshkimux I will set up some time to meet with @SKasimow to discuss and then will have a better idea of whether this is truly unblocked now. Thx!

**Comment by MarciMcGuireGCIO** [comment]
@joshkimux - @SKasimow and I are meeting tomorrow.  I will call the Forms API and update the description to include all of the impacted pages as more have been added since this was first opened in October.

**Comment by MarciMcGuireGCIO** [comment]
I've reviewed the data returned from the Forms API, and it appears that the language tag is not being set correctly.  All forms that are denoted as Spanish in their `form_name` have a value of "en" in the `language` field as shown below.

| Column1.attributes.form_name | Column1.attributes.language |
|------------------------------|-----------------------------|
| 10-0137 (espanol)            | en                          |
| 10-10EZ (esp)                | en                          |
| 21P-0510(Spanish)            | en                          |
| 21P-0512s-1(Spanish)         | en                          |
| 21P-0513-1(Spanish)          | en                          |
| 21P-0514-1(Spanish)          | en                          |
| 21P-0516-1 (Spanish)         | en                          |
| 21P-0517-1(Spanish)          | en                          |
| 21P-0518-1(Spanish)          | en                          |
| 21P-0519C-1(Spanish)         | en                          |
| 21P-0519S-1(Spanish)         | en                          |

**Comment by johnhashva** [comment]
@MarciMcGuireGCIO For posterity: I just looked at the first two in the CMS -- and both are set for Espanol as seen here.

<img width="423" alt="21P-0510" src="https://user-images.githubusercontent.com/63107147/113199924-b3a83780-9235-11eb-933c-3c768d2b8390.png">
<img width="542" alt="21P-0510 CMS Edit" src="https://user-images.githubusercontent.com/63107147/113199944-b60a9180-9235-11eb-8880-d58abe1c0940.png">

**Comment by MarciMcGuireGCIO** [comment]
@johnhashva the plot thickens.  I'll do some more research.

**Comment by MarciMcGuireGCIO** [comment]
@joshkimux we're not currently using the tag to do any language switching.  The detail pages are a mashup of boilerplate template text (currently all in English), metadata derived from the VA's forms database, and additional metadata that is manually appended to the record in the CMS.  Any place you see Spanish on a page is simply because someone copy/pasted text out of a Spanish PDF and into the metadata fields in the CMS.

**Comment by zacharymorel** [comment]
@MarciMcGuireGCIO iirc, there was a reason why things we're defaulted to EN. 
I am not sure why. @ncksllvn might have more context?

**Comment by zacharymorel** [comment]
I think it might have something to do with the **html** templates where not ready to consume a dynmaic language value for its lang html attribute.

**Comment by MarciMcGuireGCIO** [comment]
@zacharymorel that makes sense.  I know Nick is working on a platform-related thing currently, so we'll have to do some digging and try to figure it out as best we can.

**Comment by MarciMcGuireGCIO** [comment]
@zacharymorel  -- related Slack thread:

https://dsva.slack.com/archives/CUB5X5MGF/p1609864782367500?thread_ts=1609815469.366400&cid=CUB5X5MGF

**Comment by zacharymorel** [comment]
Awesome!

**Comment by MarciMcGuireGCIO** [comment]
Just to verify that I didn't import the .json into Excel incorrectly, here's the response in Postman for form **21P-0510(Spanish)**, which as shown in @johnhashva's [screen shot above](https://github.com/department-of-veterans-affairs/va.gov-team/issues/14432#issuecomment-811376075) , shows the language as Spanish.  Note `language` in the response is set to "en".

![image](https://user-images.githubusercontent.com/73354907/113205046-d3daf500-923b-11eb-9d72-247cf37787d3.png)

**Comment by zacharymorel** [comment]
We may need Mark from Lighthouse to have a look.

**Comment by MarciMcGuireGCIO** [comment]
step ahead of you! 😆 https://dsva.slack.com/archives/CUB5X5MGF/p1617221959240600

**Comment by MarciMcGuireGCIO** [comment]
Per Mark: 

> Right now the API uses langcode and not fieldVaFormLanguage

**Comment by zacharymorel** [comment]
Oh wait a minute, this 508 is for the form detail page, not the search results?

**Comment by MarciMcGuireGCIO** [comment]
@zacharymorel, correct.  The Details listed three specific form detail pages (now there are 11).

**Comment by zacharymorel** [comment]
Ah okay I can get with @youngfreezyVA and see how he is handling this so I can keep up with the standards.

**Comment by MarciMcGuireGCIO** [comment]
Per Mark Viterna with the Lighthouse team, they can change which field the API uses for the language value.  We just have to let him know when we are ready.  This could be more complicated, however, because the entirety of the content is **not** in Spanish, so language may need to be specified at the <p> level, but this is not currently supported in the CMS but could potentially be done per Kevin Walsh (although not encouraged) - see https://dsva.slack.com/archives/CUB5X5MGF/p1609815469366400

**Comment by zacharymorel** [comment]
@joshkimux so I can dynamically populate the lang="esp" attribute for each paragraph/ header/ etc that has text consumed from the CMS. However, the lang tag for the entire page, as seen here, I cannot change since most of the page is in english. Is that cool for you? 
 
 
![95137524-460fc880-0736-11eb-9747-9186320c52ac](https://user-images.githubusercontent.com/26075258/113210450-3df69880-9242-11eb-85b6-41e1eafc0f23.png)

**Comment by joshkimux** [comment]
What would be the effort required to translate all english to Spanish? (e.g. When to use this form -> Cuando usar este formulario). Ideally we'd want the html `lang` to be set to Spanish since this is a Spanish page. Otherwise:
- this may not be as discoverable to Spanish speaking veterans using search engines (from a SEO perspective)
- this won't be legible to Spanish screen reader users

Otherwise, that may work as a temporary solution @zacharymorel 

@1Copenut for awareness!

**Comment by 1Copenut** [comment]
@joshkimux thank you for the ping. There's a lot of moving parts. I can appreciate the complexity of coordinating a data API, 
CMS, and templates being hydrated with JSON. It's a lot of work to add or adapt.

Long-term I envision being able to pass a language attribute into Drupal templates while spinning up new pages. That attribute would set the `<html lang="LANG"> correctly and load Liquid templates to render the whole page in the preferred language. I'd also like to see a way to add content-block language overrides, so we could inject a block of Spanish into an English tagged page, for instance. These are lofty goals, but will help us serve the widest audience as we add more localized content.

**Comment by ncksllvn** [comment]
> @MarciMcGuireGCIO iirc, there was a reason why things we're defaulted to EN.
> I am not sure why. @ncksllvn might have more context?

Could be missing something, but the whole (modernized) website was in English up til recently, so it makes sense we would just default language to English

**Comment by MarciMcGuireGCIO** [comment]
You're right, @ncksllvn, and thanks for the context!  😄

**Comment by zacharymorel** [comment]
@1Copenut @joshkimux for today, can we solve this by adding the lang attribute to the headers and paragraph tags where needed on the templates, in this case, the va form node template? I believe @youngfreezyVA and @ncksllvn have been working on something similar?

**Comment by joshkimux** [comment]
@zacharymorel Since the `html` is already set to `en`, we would just need to add `lang="es"` attributes to the paragraph tags with Spanish content. That can work as a short term solution.

**Comment by zacharymorel** [comment]
@joshkimux Thanks for the quick feed back, I'll get with @ncksllvn and @youngfreezyVA to see what is needed on the technical side.

**Comment by zacharymorel** [comment]
Perfect, @MarciMcGuireGCIO we got a solid solution for the short term.

**Comment by MarciMcGuireGCIO** [comment]
@zacharymorel so to summarize, what we're saying is that, to fulfill the requirements of this specific defect, we will not pursue translating the template content or other metadata, but rather focus on adding the proper language attribute to those pieces that are current not in English?  Then, once the Content Localization solution is able to support, we will then look at support for templates in other languages and translating the content in those templates?

**Comment by zacharymorel** [comment]
@MarciMcGuireGCIO Yes, that's the gist of it!

**Comment by zacharymorel** [comment]
@MarciMcGuireGCIO What we will need from the CMS or whomever can edit our va forms nodes that are have Spanish, is the markdown text that is entered in for the _When to use this form_ paragraph will need the added `lang="es"` html attribute like so:

`<p lang="es">text text text</p>`

**Comment by MarciMcGuireGCIO** [comment]
> @MarciMcGuireGCIO What we will need from the CMS or whomever can edit our va forms nodes that are have Spanish, is the markdown text that is entered in for the _When to use this form_ paragraph will need the added `lang="es"` html attribute like so:
> 
> `<p lang="es">text text text</p>`

@zacharymorel - if I am interpreting this correctly, Kevin Walsh mentioned in [this Slack thread](https://dsva.slack.com/archives/CUB5X5MGF/p1609815469366400) that CMS does not currently support at the form node level.  Does that mean we can't currently request what we believe would be the interim solution?

**Comment by MarciMcGuireGCIO** [comment]
In addition to the 11 pages listed in the Problem Details above, there are three other related forms in English whose info gets pulled into the templates.  The "when to use" content for those three forms will eventually also need to be translated into Spanish.

**Comment by MarciMcGuireGCIO** [comment]
Mockup showing the source of detail page content

![image](https://user-images.githubusercontent.com/73354907/113360327-8a5fd800-9317-11eb-97e0-d4f8d1308dec.png)

**Comment by zacharymorel** [comment]
@MarciMcGuireGCIO good point, that's an issue then. I thought it was just mark up since the CMS returns a paragraph tag around the text entered in _when to use_...

**Comment by zacharymorel** [comment]
@joshkimux For the _When to use this form_ block of text (a paragraph tag with text that is generated from the CMS, tag and all), could we surround that paragraph with a div and give the div the language attribute since I cannot programmatically add the attribute to the P tag on the template?

**Comment by joshkimux** [comment]
@zacharymorel Yes, that works! The children will inherit the parent's attribute.

**Comment by zacharymorel** [comment]
@joshkimux Do you have socks yet?

**Comment by joshkimux** [comment]
@zacharymorel Not yet (I'll finally get time next week to set it up) :(
But I can hop on a screen share if you're ok with that!

**Comment by zacharymorel** [comment]
Yeah, do you have a quick second now?

**Comment by zacharymorel** [comment]
@joshkimux this is ready to view on staging!

**Comment by joshkimux** [comment]
@zacharymorel Thank you so much for the quick work! Well done 🥳 
Just one last item to correct here-- there's some hidden screen reader only text that announces "form name." Since its direct parent has the `lang` attribute set to Spanish, it's being read out in a Spanish accent.

Possible fix options:
1. translate it to "Nombre del formulario"
2. attach a `lang` attribute to English to it
3. remove `<dfn>Form title</dfn>` (This is a possibility, but I'd need more context on why Form title was added as screen reader only text to begin with before recommending removal. From a glance, this isn't providing an equivalent to anything a sighted user wouldn't have access to. If including "Form title" is important to explain that it's indeed a form title, it should exist for sighted users as well. @MarciMcGuireGCIO do you have any historical context on this decision? I'm guessing this was probably recommended by an accessibility specialist in the past for a good reason that I'm currently unaware of!)

**Comment by MarciMcGuireGCIO** [comment]
@joshkimux - thank you for the recommendations. I don't have any historical context off the top of my head, but I suspect your guess is the right answer.  I'll look at some of our closed 508 tickets for clues.

**Comment by zacharymorel** [comment]
@joshkimux  We do the same thing for the screen reader on the find forms search results page. We use an screen reader only visible node to announce "Form Name" before reading the actual form. 

Is that something we can get away from? 
I can also programmatically use the translation for "Form Name" so it's guaranteed tranlated.

**Comment by joshkimux** [comment]
@zacharymorel Yep! I'm familiar with that interaction and have written feedback on it before (I'm not sure if a ticket got written for it.) It's particularly bad on the forms search results page because "Form name:" is read out in front of _every single search result_ which causes an extremely unnecessary verbose experience.

I really want to recommend to remove it, but I'd like to understand why it was added in the first place first.

If we can't get that info in a timely fashion, programmatically translating Form name should work!

**Comment by MarciMcGuireGCIO** [comment]
@joshkimux @zacharymorel I have not been successful in finding any context around why that was added.

**Comment by zacharymorel** [comment]
@joshkimux this is on staging with translation for "form name" in es.

**Comment by joshkimux** [comment]
🔥 You're on a roll @zacharymorel , closing this one out

<img width="1274" alt="Screen Shot 2021-04-06 at 9 29 02 AM" src="https://user-images.githubusercontent.com/14154792/113718389-97374f80-96ba-11eb-8a7b-ea37c46df88c.png">

@MarciMcGuireGCIO I'm reaching out to some other people to see if they have any context on the screen reader only form name stuff. If anything it would only be a `defect-level-4` if a defect at all. I'll get back to you when I have more information to make a call on it 😄

**Comment by MarciMcGuireGCIO** [comment]
Whoo hoo!  Great job, Zach! 👏🏻.  Thanks, @joshkimux for your help, and please keep me posted on that.
